File: typescript/design-patterns/interpreter/terminals/TerminalExpressions.ts
// NumberExpression (Terminal): returns static numbers directly
class NumberExpression implements Expression {
  constructor(private value: number) {}

  interpret() {
    return this.value;
  }
}

// VariableExpression (Terminal): looks up variables inside the context map
class VariableExpression implements Expression {
  constructor(private name: string) {}

  interpret(context: Map<string, number>): number {
    if (!context.has(this.name)) {
      throw new Error(\

File: python/design-patterns/interpreter/terminals/terminal_expressions.py
class NumberExpression(Expression):
    def __init__(self, value: float):
        self.value = value
    def interpret(self, context): return self.value

class VariableExpression(Expression):
    def __init__(self, name: str):
        self.name = name
    def interpret(self, context):
        return context[self.name]